artiq/applets/big_number.py

```diff
@@ -26,54 +27,76 @@ def __init__(self, args, ctl):
         QtWidgets.QStackedWidget.__init__(self)
         self.dataset_name = args.dataset
         self.ctl = ctl
+        self.value = None
+        self.metadata = {}
 
-        self.lcd_widget = QResponsiveLCDNumber()
-        self.lcd_widget.setDigitCount(args.digit_count)
-        self.lcd_widget.doubleClicked.connect(self.start_edit)
-        self.addWidget(self.lcd_widget)
+        font = QtGui.QFont()
+        font.setPointSize(60)
+
+        self.display_widget = QResponsiveLabel()
+        self.display_widget.setFont(font)
+        self.display_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.display_widget.doubleClicked.connect(self.start_edit)
+        self.addWidget(self.display_widget)
 
         self.edit_widget = QCancellableLineEdit()
-        self.edit_widget.setValidator(QtGui.QDoubleValidator())
-        self.edit_widget.setAlignment(QtCore.Qt.AlignRight)
-        self.edit_widget.editCancelled.connect(self.cancel_edit)
+        self.edit_widget.setFont(font)
+        self.edit_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.edit_widget.editCancelled.connect(
+            lambda: self.setCurrentWidget(self.display_widget))
         self.edit_widget.returnPressed.connect(self.confirm_edit)
         self.addWidget(self.edit_widget)
 
-        font = QtGui.QFont()
-        font.setPointSize(60)
-        self.edit_widget.setFont(font)
-
-        self.setCurrentWidget(self.lcd_widget)
+        self.setCurrentWidget(self.display_widget)
 
     def start_edit(self):
-        # QLCDNumber value property contains the value of zero
-        # if the displayed value is not a number.
-        self.edit_widget.setText(str(self.lcd_widget.value()))
+        if self.value is None:
+            self.edit_widget.setText("---")
+        else:
+            scale = scale_from_metadata(self.metadata)
+            value = self.value/scale
+            unit = self.metadata.get("unit", "")
+            self.edit_widget.setText(f"{value} {unit}".strip())
         self.edit_widget.selectAll()
         self.edit_widget.setFocus()
         self.setCurrentWidget(self.edit_widget)
 
     def confirm_edit(self):
-        value = float(self.edit_widget.text())
-        self.ctl.set_dataset(self.dataset_name, value)
-        self.setCurrentWidget(self.lcd_widget)
-
-    def cancel_edit(self):
-        self.setCurrentWidget(self.lcd_widget)
+        entry = self.edit_widget.text().split()
+        if not entry:
+            self.reset_display()
+            return
+        try:
+            value = float(entry[0])
+        except ValueError:
+            self.reset_display()
+            return
+        unit = entry[1] if len(entry) > 1 else ""
+        scale = scale_from_metadata({"unit": unit})
+        value = value*scale
```

Comment:
We override the scaling here as an explicit 'scale' parameter has not been considered in the widget, so anytime it is edited it will use the default unit system for overriding. This is dodgy in my opinion, would be good to consider a solution to this @fsagbuya.

Reply:
Line 77 can probably be better rewritten using the 'getattr(units, unit)' way used in many other places. The 'scale_from_metadata' helper is used for resolving conflicts between the default unit scaling and a predefined scale.